python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -0,0 +1,323 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from abc import ABC, abstractmethod
+
+import cudf
+
+from cuspatial.core._column.geocolumn import GeoColumn
+from cuspatial.core.binpreds.contains import contains_properly
+from cuspatial.utils.column_utils import (
+    contains_only_linestrings,
+    contains_only_multipoints,
+    contains_only_polygons,
+    has_same_geometry,
+)
+
+
+class BinaryPredicate(ABC):
+    @abstractmethod
+    def preprocess(self, lhs, rhs):
+        """Preprocess the input data for the binary operation. This method
+        should be implemented by subclasses. Preprocess and postprocess are
+        used to implement the discrete math rules of the binary operations.
+
+        Notes
+        -----
+        Should only be called once.
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to perform.
+        lhs : GeoSeries
+            The left-hand-side of the GeoSeries-level binary operation.
+        rhs : GeoSeries
+            The right-hand-side of the GeoSeries-level binary operation.
+
+        Returns
+        -------
+        GeoSeries
+            The left-hand-side of the internal binary operation, may be
+            reordered.
+        GeoSeries
+            The right-hand-side of the internal binary operation, may be
+            reordered.
+        """
+        return (lhs, rhs, cudf.RangeIndex(len(rhs)))
+
+    @abstractmethod
+    def postprocess(self, point_indices, point_result):
+        """Postprocess the output data for the binary operation. This method
+        should be implemented by subclasses.
+
+        Postprocess converts the raw results of the binary operation into
+        the final result. This is where the discrete math rules are applied.
```

Comment:
What discrete math rules do you mean?

Reply:
I rewrote the sentence as: "At this step the results for none, any, and all
        are applied to the result of the equals, intersects, and
        point-in-polygon operations."